server/src/leaderboard.ts

```diff
@@ -0,0 +1,328 @@
+  daily: Record<number, DailyLeaderboard>
+  allTime: {
+    mostSolved: SolvedScore[]
+    currentStreak: StreakScore[]
+    fastest: SpeedScore[]
+  }
+}
+
+export const leaderboard: Leaderboard = {
+  daily: {},
+  allTime: {
+    mostSolved: [],
+    currentStreak: [],
+    fastest: []
+  }
+}
+
+function beginningOfTodayInUTCSeconds(): number {
+  const now = new Date()
+  const startOfTodayUTC = new Date(
+    Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate())
+  )
+  return Math.floor(startOfTodayUTC.getTime() / 1000)
+}
+
+async function calculateDailyLeaderboard(day: number, chain: Chain) {
+  const result = await db.query(`
+  WITH fastest_times AS (
+    SELECT 
+        problem_id,
+        level,
+        MIN(ticks_in_this_match) AS fastest_time
+    FROM 
+        solver_solved
+    WHERE 
+        level IN (1, 2, 3)
+        AND day = ${day}
+        AND src_name = '${chain}'
+    GROUP BY 
+        problem_id, level
+  ),
+  cumulative_times AS (
+      SELECT 
+          problem_id,
+          SUM(ticks_in_this_match) AS total_time
+      FROM 
+          solver_solved
+      WHERE 
+          day = ${day}
+          AND src_name = '${chain}'
+      GROUP BY 
+          problem_id
+      HAVING 
+          COUNT(level) = 3
+  ),
+  ranked_fastest_times AS (
+      SELECT 
+          problem_id,
+          level,
+          fastest_time,
+          ROW_NUMBER() OVER (PARTITION BY level ORDER BY fastest_time) AS rank
+      FROM 
+          fastest_times
+  ),
+  ranked_cumulative_times AS (
+      SELECT 
+          problem_id,
+          total_time,
+          ROW_NUMBER() OVER (ORDER BY total_time) AS rank
+      FROM 
+          cumulative_times
+  ),
+  leaderboard AS (
+    SELECT 
+        'Level 1' AS category,
+        problem_id,
+        fastest_time AS time
+    FROM 
+        ranked_fastest_times
+    WHERE 
+        level = 1 AND rank <= 3
+    UNION ALL
+    SELECT 
+        'Level 2' AS category,
+        problem_id,
+        fastest_time AS time
+    FROM 
+        ranked_fastest_times
+    WHERE 
+        level = 2 AND rank <= 3
+    UNION ALL
+    SELECT 
+        'Level 3' AS category,
+        problem_id,
+        fastest_time AS time
+    FROM 
+        ranked_fastest_times
+    WHERE 
+        level = 3 AND rank <= 3
+    UNION ALL
+    SELECT 
+        'Cumulative' AS category,
+        problem_id,
+        total_time AS time
+    FROM 
+        ranked_cumulative_times
+    WHERE 
+        rank <= 3
+    ORDER BY 
+        category, time
+  ),
+  latest_transactions AS (
+    SELECT
+        token_id,
+        "to",
+        ROW_NUMBER() OVER (PARTITION BY token_id ORDER BY block_num DESC, tx_idx DESC, log_idx DESC) AS rn
+    FROM
+      problems_transfer
+    WHERE
+      token_id IN (SELECT problem_id FROM leaderboard)
+      AND src_name = '${chain}'
+  ),
+  current_owners AS (
+      SELECT
+          token_id,
+          concat('0x', encode("to", 'hex')) as owner
```

Comment:
ah i see you're adding it back here now, ignore other comment about this

Reply:
I add it when it comes out of the DB and remove when it goes in because that's how the addresses get stored by shovel as `bytea`